File: arduinobot_ws/src/arduinobot_firmware/arduinobot_firmware/simple_serial_receiver.py
#!/usr/bin/env python3
"""
Module providing a ROS 2 node to receive serial messages from Arduino.
"""
import serial
import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class SimpleSerialReceiver(Node):
    """
    ROS 2 Node that reads lines from a serial port and publishes them as Strings.

    This node polls the serial interface at a fixed frequency, decodes the incoming
    bytes as UTF-8, and broadcasts the data on the 'serial_receiver' topic.
    """
    def __init__(self):
        """
        Initialize the SimpleSerialReceiver node.

        Declares 'port' and 'baudrate' parameters, opens the serial connection,
        and initializes the publisher and a timer to check for incoming data.
        """
        super().__init__("simple_serial_receiver")

        self.declare_parameter("port", "/dev/ttyUSB0")
        self.declare_parameter("baudrate", 115200)

        self.port_ = self.get_parameter("port").value
        self.baudrate_ = self.get_parameter("baudrate").value

        self.pub_ = self.create_publisher(String, "serial_receiver", 10)
        self.arduino_ = serial.Serial(port=self.port_, baudrate=self.baudrate_, timeout=0.1)

        self.frequency_ = 0.01
        self.timer_ = self.create_timer(self.frequency_, self.timerCallback)

    def timerCallback(self):
        """
        Timer callback that executes at 100Hz to read data from the serial port.

        Decodes the serial line as utf-8 and publishes it to the 'serial_receiver' topic.
        """
        if rclpy.ok() and self.arduino_.is_open:
            data = self.arduino_.readline()

            try:
                data.decode("utf-8")
            except:  # noqa: E722
                return

            msg = String()
            msg.data = str(data)
            self.pub_.publish(msg)


def main():
    rclpy.init()

    simple_serial_receiver = SimpleSerialReceiver()
    rclpy.spin(simple_serial_receiver)
    
    simple_serial_receiver.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: arduinobot_ws/src/arduinobot_firmware/src/simple_serial_receiver.cpp
/**
 * @file simple_serial_receiver.cpp
 * @brief ROS 2 Node that receives serial data from Arduino and publishes it.
 *
 * This node opens a serial connection to the designated port, polls the serial
 * port at 100 Hz, reads lines of text, and publishes them as ROS 2 String messages.
 */

#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>

#include <chrono>

#include <libserial/SerialPort.h>

using namespace std::chrono_literals;

/**
 * @class SimpleSerialReceiver
 * @brief Node that listens to serial port signals and broadcasts them onto a ROS 2 topic.
 */
class SimpleSerialReceiver : public rclcpp::Node
{
public:
  /**
   * @brief Construct a new Simple Serial Receiver object.
   *
   * Declares parameters, opens the serial connection at 115200 baud,
   * sets up the publisher, and starts a wall timer to poll for data.
   */
  SimpleSerialReceiver() : Node("simple_serial_receiver")
  {
    declare_parameter<std::string>("port", "/dev/ttyUSB0");

    port_ = get_parameter("port").as_string();

    arduino_.Open(port_);
    arduino_.SetBaudRate(LibSerial::BaudRate::BAUD_115200);

    pub_ = create_publisher<std_msgs::msg::String>("serial_receiver", 10);
    timer_ = create_wall_timer(0.01s, std::bind(&SimpleSerialReceiver::timerCallback, this));
  }

  /**
   * @brief Destroy the Simple Serial Receiver object.
   *
   * Ensures the serial interface connection is closed correctly.
   */
  ~SimpleSerialReceiver()
  {
    arduino_.Close();
  }

  /**
   * @brief Callback function executed by the timer to poll the serial device.
   *
   * Verifies if any bytes are available to read. If yes, reads the line
   * and publishes it on the 'serial_receiver' topic.
   */
  void timerCallback()
  {
    if(rclcpp::ok() && arduino_.IsDataAvailable())
    {
      auto message = std_msgs::msg::String();
      arduino_.ReadLine(message.data);
      pub_->publish(message);
    }
  }

private:
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_;
  rclcpp::TimerBase::SharedPtr timer_;
  std::string port_;
  LibSerial::SerialPort arduino_;
};


int main(int argc, char* argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SimpleSerialReceiver>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}